### PR TITLE
Move System File Table memory location

### LINKED ISSUE #3668: Dunkle Schatten 2 freezes dosbox
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

0.81.1

### What Operating System are you using?

Windows 10

### If Other OS, please describe

_No response_

### Relevant hardware info

i73770k, GTX 1070

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

The **_free advertising adventure_**  "Dunkle Schatten 2" freezes as soon as you try to start it. It works with dosbox 0.74-3.

Download: http://www.mediafire.com/download/z1z9zd139hqd235/Dunkle+Schatten+2.7z

### Steps to reproduce the behaviour.

Explain how to reproduce
1.
mount c ..\SCHATTEN2
C:
DS2.BAT

2. dosbox-staging freezes

### Download URL of affected game or software

_No response_

### Your configuration

```ini
Default 0.81.1 settings
```

### Provide a Log

_No response_

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

### PR DESCRIPTION
**Pull request by weirddan455**
# Description

The System File Table (SFT) was being faked and did not have enough memory allocated for it. This caused issues with Dunkle Schatten 2.

The game was making an INT 2F call that requests a pointer to an SFT entry. This interrupt handler writes into SFT memory which was overlapping with the DOS Device Chain, causing a hang.

To fix, actually allocate enough memory to store 100 file handles and set the pointer to firstFileTable to reflect the new location.

## Related issues

Fixes #3668

# Manual testing

The game doesn't crash anymore :smile: 

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: src/dos/dos_classes.cpp [hunk 1]
```diff
@@ -118,17 +118,23 @@ void DOS_InfoBlock::SetLocation(uint16_t segment)
...
+	// This actually gets writen to by DOS_MultiplexFunctions() so we must allocate enough memory
+	constexpr uint32_t EndPointer = 0xffffffff;
+	constexpr uint16_t NextTableOffset = 0x0;
+	constexpr uint16_t NumberOfFilesOffset = 0x04;
+
+	constexpr int BytesPerFile = 59;
+	constexpr int NumberOfFiles = 100;
+	constexpr int BytesPerPage = 16;
+	constexpr int TotalBytes = BytesPerFile * NumberOfFiles;
+	constexpr int NumberOfPages = (TotalBytes / BytesPerPage) + std::min(TotalBytes % BytesPerPage, 1);
```
Comment (johnnovak) [inline src/dos/dos_classes.cpp:130]: Minor: NumFiles, NumPages ?

**Review by johnnovak — APPROVED** [review]
Nice one @weirddan455. Approving as I have no idea how to test this (apart from this game), so I'll trust you on this one.

Can you think of any other areas where this could introduce problems or regressions? If we're closer to "real DOS" with this change, I guess not... but you never know 😅

I'd suggest to maybe regression test it a bit more before merging.

**Comment by weirddan455** [comment]
The `shell_cmds` test is failing so I'm not merging yet.  Something in there must be allocating a lot of memory because it's throwing a DOS out of memory error and aborting.

**Comment by codengine** [comment]
> Nice one @weirddan455. Approving as I have no idea how to test this (apart from this game), so I'll trust you on this one.
> 
> Can you think of any other areas where this could introduce problems or regressions? If we're closer to "real DOS" with this change, I guess not... but you never know 😅
> 
> I'd suggest to maybe regression test it a bit more before merging.

Regarding testing, we will take care at least for Dunkle Schatten 2 and Mad TV 2 whose ticket can then also be closed if it works.

File: src/dos/dos_classes.cpp (hunk 1 above)
Comment (weirddan455) [reply src/dos/dos_classes.cpp]: Renamed

**Comment by weirddan455** [comment]
Just pushed a commit that fixes the `shell_cmds` test on my local machine.  The internal DOS tables memory was allocated with every test and never being freed so it would pass a few tests and then run out of memory.  It's a simple "bump allocator" I think you could call it so freeing is just resetting a variable back to default.

**Comment by weirddan455** [comment]
memory

**Comment by weirddan455** [comment]
I put a big ole `LOG_ERR` at the top of the multiplex function that writes to the SFT and tested the following games.  They all work and none of them call this function.

https://github.com/dosbox-staging/dosbox-staging/blob/5c0e311d4de8d2fc39b601a58a1212fa68c42777/src/dos/dos_misc.cpp#L64-L72

- Doom
- Quake
- Secret of Monkey Island 2
- Space Quest 1 (tested with `machine = tandy`, `machine = pcjr`, and `machine = svga_s3`)
- Space Quest 3
- King's Quest 1 (pcjr booter version)
- Sim City 2000
- Tyrian

**Comment by johnnovak** [comment]
Okay, @weirddan455, here's my part. All looking good so far 👍🏻 

**Apps**

- DOS Navigator
- DPaint IIe
- FastTracker II
- QPV
- RealSpec (ZX Spectrum emulator)

**Games**

With the machine types I tested with in parenthesis:

- Adventures in Serenia (PCjr, booter)
- Ancient Art of War (CGA)
- Below the Root (Tandy)
- Bruce Lee (PCjr, CGA)
- Demon's Forge (Tandy, booter)
- Life and Death (CGA)
- Railroad Tycoon Deluxe (VGA)
- Starflight (VGA)
- Total Eclipse (Hercules, PCjr, Tandy, CGA, EGA, VGA)
- Wasteland (VGA)
- Zak McKracken and the Alien Mindbenders Enhanced (Tandy)

**Extra**

- Bunch of random demoscene stuff (intros, demos, etc.)

**Comment by Max833** [comment]
Thank you very much for investigating and resolving this issue! :)
According to someone who is currently testing a lot of games, there are issues with Windows 3.1 games now with this commit (they don't start).

**Comment by weirddan455** [comment]
> Thank you very much for investigating and resolving this issue! :) According to someone who is currently testing a lot of games, there are issues with Windows 3.1 games now with this commit (they don't start).

Which games?  And can you confirm it's actually a regression by this commit (aka. they start on `main`)

**Comment by johnnovak** [comment]
> > Thank you very much for investigating and resolving this issue! :) According to someone who is currently testing a lot of games, there are issues with Windows 3.1 games now with this commit (they don't start).
> 
> Which games? And can you confirm it's actually a regression by this commit (aka. they start on `main`)

Confirmed, Windows for Workgroups 3.11 itself fails to start. Just a black screen after

**Comment by MasterO2** [comment]
So essentially, the System File Table needs to be in a memory location that does not tread upon other functions already mapped to memory?

**Comment by codengine** [comment]
> So essentially, the System File Table needs to be in a memory location that does not tread upon other functions already mapped to memory?

Dunkle Schatten 2 uses INT2F to access files and while doing so steps on memory areas that are mapped to the device chain. So yeah it overlaps.

**Comment by weirddan455** [comment]
Windows 3.11 starts up fine for me on this branch.  Is Windows for Workgroups something different and the problem only exists there?

image0001

**Comment by codengine** [comment]
> Windows 3.11 starts up fine for me on this branch. Is Windows for Workgroups something different and the problem only exists there?

Yes it has advanced networking. Probably because of that the device chain was added at the first place.

**Comment by johnnovak** [comment]
@weirddan455 Yeah, it's different and for reasons not completely clear to me, WfW is recommended in many DOSBox guides.

I've done my Windows for Workgroups 3.11 setup using our guide:

https://github.com/dosbox-staging/dosbox-staging/wiki/Windows-3.1x

Happy to send you a ZIP of my setup if that helps.

**Comment by weirddan455** [comment]
Ok I've reproduced the bug.  It seems Windows for Workgroups requires a 2nd SFT table to be chained onto the first which I had removed in my PR.  It doesn't seem to care where it is located in memory, it just has to exist somewhere.  Windows never calls into that multiplex function either so it must be reading the memory directly or something.

I also just realized I could get away with allocating less memory.  The multiplex function has a check where it's limited to 16 files so there's no need to allocate for 100.

Side-note (this was unchanged in my PR) but this bit of code is required or Windows aborts with a message "Invalid DOS version".  All this is doing is writing the string "CON" to memory.  This is nowhere near in memory to where I've relocated the SFT table but it still must exist.  Makes me think even more that Windows is directly scanning the memory.

https://github.com/dosbox-staging/dosbox-staging/blob/3db71d5a6ba5d23a827448480d316d464bfa219a/src/dos/dos_tables.cpp#L85-L89

I'll update this PR with a fix soon.

**Comment by interloper98** [comment]
An ex-DOS MS-DOS 4.x developer has this story that you might find interesting, @weirddan455 :

> when Windows started up, it increased the size of MS-DOS’s internal file table (the SFT, that’s the table that was created by the FILES= line in config.sys).  It did that to allow more than 20 files to be opened on the windows system (a highly desirable goal for a multi-tasking operating system).

> But it did that by using an undocumented API call, which returned a pointer to a set of “interesting” pointers in MS-DOS. It then indexed a known offset relative to that pointer, and replaced the value of the master SFT table with its own version of the SFT.

> When I was working on MS-DOS 4.0, we needed to support Windows.  Well, it was relatively easy to guarantee that our SFT was at the location that Windows was expecting.  But the problem was that the MS-DOS 4.0 SFT was 2 bytes larger than the MS-DOS 3.1 SFT.

> In order to get Windows to work, I had to change the DOS loader to detect when win.com was being loaded, and if it was being loaded, I looked at the code at an offset relative to the base code segment, and if it was a “MOV” instruction, and the amount being moved was the old size of the SFT, I patched the instruction in memory to reflect the new size of the SFT!  Yup, MS-DOS 4.0 patched the running windows binary to make sure Windows would still continue to work.

https://learn.microsoft.com/en-us/archive/blogs/larryosterman/aardvarks-in-your-code

**Comment by johnnovak** [comment]
Very nice quick fix @weirddan455. My WfW 3.11 installation is back in business 🎉 

image

**Comment by Max833** [comment]
Looking good! It doesn't work with a fresh installation (https://archive.org/details/dunkle-schatten-1-2, see video), but the reason for that is DOS4GW (same issue with DOSBox 0.74). It's working fine when you replace the extender (dos32a).

https://streamable.com/lhlif0

**Comment by MasterO2** [comment]
On a semi-related note, I have added `Dunkel Schatten 2` to https://github.com/dosbox-staging/dosbox-staging/wiki/DOS-32A-compatibility-list.

**Comment by johnnovak** [comment]
Can we merge this @codengine @weirddan455 @Max833 @interloper98? Or do you guys wanna test a few more things? It's looking good from my end.

**Comment by interloper98** [comment]
Been using this CI build in WfW 3.11 + MS Office, Tie Fighter, Falcon 3.0, and Command and Conquer Red Alert: working good in all of them!

**Comment by codengine** [comment]
I haven't heard any issues so far. @Max833 did you retest MAD TV 2?

**Comment by Max833** [comment]
> Can we merge this @codengine @weirddan455 @Max833 @interloper98? Or do you guys wanna test a few more things? It's looking good from my end.

I couldn't find any issues (tested around 15 games). Looking good! =)

**Comment by johnnovak** [comment]
Thanks everybody for testing, appreciated! 👍🏻 Merging.

**Comment by joncampbell123** [comment]
DOSBox-X developer here: You may be interested in recent changes to SFT tables in DOSBox-X that eliminate memory overlapping issues and arrange SFTs in a way that matches how MS-DOS does it. Windows 3.1 is happy again.

**Comment by johnnovak** [comment]
> DOSBox-X developer here: You may be interested in recent changes to SFT tables in DOSBox-X that eliminate memory overlapping issues and arrange SFTs in a way that matches how MS-DOS does it. Windows 3.1 is happy again.

Thanks @joncampbell123. Pinging @weirddan455.

**Pull request merged**